### PR TITLE
refactor: split subscribe ticker and candle channels

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Split subscription of ticker and candle channels for the following providers: Okx, Binance, Huobi, and Kraken
- Add error check for `websocket.CloseAbnormalClosure` to Kraken when EOF occurs to avoid having the following 
image

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: price-feeder/oracle/provider/binance.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func NewBinanceProvider(ctx context.Context, logger zerolog.Logger, pairs ...typ
 		subscribedPairs: map[string]types.CurrencyPair{},
 	}
 
-	if err := provider.SubscribeTickers(pairs...); err != nil {
+	if err := provider.SubscribeCurrencyPais(pairs...); err != nil {
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:94]:
```suggestion
	if err := provider.SubscribeCurrencyPairs(pairs...); err != nil {
```

File: price-feeder/oracle/provider/binance.go [hunk 2]
```diff
@@ -152,6 +132,55 @@ func (p *BinanceProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[stri
 	return candlePrices, nil
 }
 
+// SubscribeCurrencyPais subscribe all currency pairs into ticker and candle channels.
+func (p *BinanceProvider) SubscribeCurrencyPais(cps ...types.CurrencyPair) error {
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:136]: Same as above : 
```suggestion
func (p *BinanceProvider) SubscribeCurrencyPairs(cps ...types.CurrencyPair) error {
```

Also, if we expect all providers to have the `SubscribeCurrencyPairs` function implemented going forward, we should add it to the interface : 
```golang
// Provider defines an interface an exchange price provider must implement.
type Provider interface {
	GetTickerPrices(...types.CurrencyPair) (map[string]TickerPrice, error)
	GetCandlePrices(...types.CurrencyPair) (map[string][]CandlePrice, error)
	SubscribeCurrencyPairs(...types.CurrencyPair) error
}
```

File: price-feeder/oracle/provider/binance.go [hunk 3]
```diff
@@ -152,6 +132,55 @@ func (p *BinanceProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[stri
...
+	if err := p.subscribeChannels(cps...); err != nil {
+		return err
+	}
+
+	p.setSubscribedPairs(cps...)
+	return nil
+}
+
+// subscribeChannels subscribe all currency pairs into ticker and candle channels.
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:145]:
```suggestion
// subscribeChannels subscribe to the ticker and candle channels for all currency pairs.
```

File: price-feeder/oracle/provider/binance.go [hunk 4]
```diff
@@ -152,6 +132,55 @@ func (p *BinanceProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[stri
...
+func (p *BinanceProvider) subscribeChannels(cps ...types.CurrencyPair) error {
+	if err := p.subscribeTickers(cps...); err != nil {
+		return err
+	}
+
+	return p.subscribeCandles(cps...)
+}
+
+// subscribeTickers subscribe all currency pairs into ticker channel.
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:154]:
```suggestion
// subscribeTickers subscribe to the ticker channel for all currency pairs.
```

File: price-feeder/oracle/provider/binance.go [hunk 5]
```diff
@@ -152,6 +132,55 @@ func (p *BinanceProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[stri
...
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToBinanceTickerPair(cp)
+	}
+
+	return p.subscribePairs(pairs...)
+}
+
+// subscribeCandles subscribe all currency pairs into candle channel.
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:165]: Same suggestion as above : 
```suggestion
// subscribeCandles subscribe to the candle channel for all currency pairs.
```

File: price-feeder/oracle/provider/binance.go [hunk 6]
```diff
@@ -152,6 +132,55 @@ func (p *BinanceProvider) GetCandlePrices(pairs ...types.CurrencyPair) (map[stri
...
+	pairs := make([]string, len(cps))
+
+	for i, cp := range cps {
+		pairs[i] = currencyPairToBinanceCandlePair(cp)
+	}
+
+	return p.subscribePairs(pairs...)
+}
+
+// subscribedPairsToSlice returns the map of subscribed pairs as slice
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/binance.go:176]:
```suggestion
// subscribedPairsToSlice returns the map of subscribed pairs as a slice.
```

File: price-feeder/oracle/provider/provider.go [hunk 7]
```diff
@@ -91,3 +91,16 @@ func newCandlePrice(provider, symbol, lastPrice, volume string, timeStamp int64)
...
+func subscribedPairsToSlice(subscribedPairs map[string]types.CurrencyPair) []types.CurrencyPair {
+	currencyPairs := make([]types.CurrencyPair, len(subscribedPairs))
+
+	iterator := 0
+	for _, cp := range subscribedPairs {
+		currencyPairs[iterator] = cp
+		iterator++
+	}
+
+	return currencyPairs
```
Comment (adamewozniak) [inline price-feeder/oracle/provider/provider.go:105]:
```suggestion
	var currencyPairs []types.CurrencyPair
	
	for _, cp := range subscribedPairs {
		currencyPairs = append(currencyPairs, cp)
	}

	return currencyPairs
```

**Review by adamewozniak — COMMENTED** [review]
Nice! Just left a few minor suggestions - let's merge this before #611

File: price-feeder/oracle/provider/provider.go [hunk 8]
```diff
@@ -91,3 +91,14 @@ func newCandlePrice(provider, symbol, lastPrice, volume string, timeStamp int64)
 func PastUnixTime(t time.Duration) int64 {
 	return time.Now().Add(t*-1).Unix() * int64(time.Second/time.Millisecond)
 }
+
+// mapPairsToSlice returns the map of currency pairs as slice.
+func mapPairsToSlice(mapPairs map[string]types.CurrencyPair) []types.CurrencyPair {
+	var currencyPairs []types.CurrencyPair
```
Comment (alexanderbez) [inline price-feeder/oracle/provider/provider.go:97]:
```suggestion
	currencyPairs := make([]types.CurrencyPair, len(mapPairs))
```

Then you can avoid the `append` and use an index :)

**Review by alexanderbez — COMMENTED** [review]
LGTM. Just one minor nit

File: price-feeder/oracle/provider/provider.go (hunk 8 above)
Comment (adamewozniak) [reply price-feeder/oracle/provider/provider.go:97]: This is my bad - I'm used to prioritizing less lines of code, which is not golangish

**Comment by RafilxTenfen** [comment]
> LGTM - what do you think about adding `SubscribeCurrencyPairs` to the interface?

I thought about that, but what about Osmosis? who doesn't have websocket 
@adamewozniak

**Review by alexanderbez — APPROVED** [review]

**Comment by adamewozniak** [comment]
> > LGTM - what do you think about adding `SubscribeCurrencyPairs` to the interface?
> 
> I thought about that, but what about Osmosis? who doesn't have websocket @adamewozniak

Could put in a no-op, not sure what best practice is to do here. Imo it might be easier for library users (peggo) to assume that the method will be implemented on all providers. 
cc : @alexanderbez

**Comment by alexanderbez** [comment]
I imagine we could add `SubscribeCurrencyPairs` to our interface. In osmosis and others that dont provide websockets, can we have them safely perform a no-op?

**Comment by alexanderbez** [comment]
Shall we merge this PR or do we want to make further tweaks?

**Comment by adamewozniak** [comment]
> Shall we merge this PR or do we want to make further tweaks?

imo we can merge, I can update the interface in #611 later today
cc : @RafilxTenfen

**Pull request merged**